src/main/scala/org/constellation/primitives/EdgeDAO.scala

```diff
@@ -1,15 +1,178 @@
 package org.constellation.primitives
 
-import java.util.concurrent.Executors
+import java.util.concurrent.{Executors, TimeUnit}
 
-import org.constellation.ProcessingConfig
+import akka.util.Timeout
+import org.constellation.consensus.EdgeProcessor.acceptCheckpoint
+import org.constellation.consensus.{Snapshot, SnapshotInfo, TipData}
 import org.constellation.primitives.Schema._
+import org.constellation.{DAO, ProcessingConfig}
 
 import scala.collection.concurrent.TrieMap
 import scala.concurrent.{ExecutionContext, ExecutionContextExecutor}
+import scala.util.Random
+
+
+class ThreadSafeTXMemPool() {
+
+  private var transactions = Seq[Transaction]()
+
+  def pull(minCount: Int): Option[Seq[Transaction]] = this.synchronized{
+    if (transactions.size > minCount) {
+      val (left, right) = transactions.splitAt(minCount)
+      transactions = right
+      Some(left)
+    } else None
+  }
+
+  def put(transaction: Transaction): Boolean = this.synchronized{
+    if (!transactions.contains(transaction)) {
+      if (transactions.size < 1000) {
+        transactions :+= transaction
+      }
+      true
+    } else false
+  }
+
+  def unsafeCount: Int = transactions.size
```

Comment:
why is this unsafe?

Reply:
Because if a thread expects a condition on it (like only do something if greater or less than a count) it won't actually reflect the real count because other threads may be modifying it. It will be fine for metrics calls since those don't matter - which is what it's used for.

Reply:
that's true -- but that would be true of anything that changes. count only exists at a given point in time. Anyways -- it's fine